src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs

```diff
@@ -55,34 +50,13 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
                 ? GetPackageSources(document)
                 : ImmutableArray<PackageSource>.Empty;
 
-            // We might have multiple different diagnostics covering the same span.  Have to
-            // process them all as we might produce different fixes for each diagnostic.
-
-            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
-            var placeSystemNamespaceFirst = documentOptions.GetOption(GenerationOptions.PlaceSystemNamespaceFirst);
+            var fixesForDiagnostic = await addImportService
+                .GetFixesForDiagnosticsAsync(document, span, diagnostics, symbolSearchService, searchReferenceAssemblies, packageSources, cancellationToken).ConfigureAwait(false);
 
-            foreach (var diagnostic in context.Diagnostics)
+            foreach (var (diagnostic, fixes) in fixesForDiagnostic)
             {
-                var fixes = await addImportService.GetFixesAsync(
-                    document, span, diagnostic.Id, placeSystemNamespaceFirst,
-                    symbolSearchService, searchReferenceAssemblies, 
-                    packageSources, cancellationToken).ConfigureAwait(false);
-
-                var codeActions = ArrayBuilder<CodeAction>.GetInstance();
-
-                foreach (var fix in fixes)
-                {
-                    var codeAction = TryCreateCodeAction(document, fix);
-                    codeActions.AddIfNotNull(codeAction);
-
-                    if (codeActions.Count >= MaxResults)
-                    {
-                        break;
-                    }
-                }
-
+                var codeActions = addImportService.GetCodeActionsForFixes(document, fixes, GetPackageInstallerService(document));
```

Comment:
is GetPackageInstallerService cheap to get? otherwise, outside of loop? since it will always give back same one?